### PR TITLE
ocamlPackages.jupyter: init at f886e34

### PR DESCRIPTION
**Pull request by kwshi**
###### Motivation for this change

The `jupyter` NixOS module allows [creation of custom kernels](https://nixos.org/nixos/options.html#services.jupyter.kernels).  [OCaml](https://ocaml.org/) is a popular functional programming language and [OCaml Jupyter](https://github.com/akabe/ocaml-jupyter) is a popular OCaml kernel for the Jupyter notebook server.  This PR adds the OCaml Jupyter kernel to `ocamlPackages`, making it possible to create custom OCaml kernels for NixOS-run Jupyter servers.

###### Things done

(This is my first attempt at contributing to nixpkgs, so please let me know what I can/should be doing better!)

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vbgl** [comment]
Thanks for your contribution!

Can you please quickly explain how to try this out?

The program `ocaml-jupyter-opam-genspec` seems completely broken. I guess it is better not to install it.

Also, It would be nice if you could find a better version numbers (increasing numbers are preferred).

**Comment by kwshi** [comment]
@vbgl AFAIK `ocaml-jupyter-opam-genspec` is meant for those who use [OPAM](http://opam.ocaml.org/) to manage their OCaml install/packages.  Its only role is, as a convenience, to auto-generate the [Jupyter kernelspec files](https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs) using OPAM variables to point to the correct paths where the kernel binary is installed.

For those who use Nix to manage their OCaml packages, that program is not relevant since (1) it tries to ask OPAM for its install path, and OPAM may very well not be installed; (2) whatever paths it generates are irrelevant, since the actual binary would be installed in the Nix store, not the OPAM store; (3) NixOS users won't use the kernelspec files produced by the genspec program anyway but rather configure it through the NixOS module system.

Following the example given in the NixOS module docs, here is a sample configuration (in configuration.nix) that would be used to setup an OCaml kernel:
```
{
  ocaml = {
    displayName = "OCaml";
    argv = [
      "${ocamlPackages.jupyter}/bin/ocaml-jupyter-kernel"
      "--connection_file"
      "{connection_file}"
    ];
    language = "ocaml";
  };
}
```

Finally, regarding version numbers: previous versions of the OCaml Jupyter package were not build-able on NixOS without OPAM, so I had opened an issue and the author had recently fixed it to make it friendly to non-OPAM users.  There has not been another actual "release" since then, so instead of a proper version I picked the latest commit ID to make it work.  When the author of that package does make a release version, then I will update the version here to match.

**Comment by bcdarwin** [comment]
Hi @kwshi, it looks like upstream has made a new release.  Would you consider updating your PR?

File: pkgs/development/ocaml-modules/jupyter/default.nix [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    opam
+  ];
+
+  meta = {
+    homepage = https://github.com/akabe/ocaml-jupyter;
+    description = "An OCaml kernel for Jupyter (IPython) notebook";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ kwshi ];
+  };
+}
```
Comment (ulrikstrid) [inline pkgs/development/ocaml-modules/jupyter/default.nix:55]:
```suggestion
{ lib
, fetchurl
, buildDunePackage
, cppo
, base
, uuidm
, base64
, lwt
, lwt_ppx
, logs
, stdint
, zmq
, zmq-lwt
, yojson
, ppx_yojson_conv
, ppx_deriving
, cryptokit
, opam
, ounit2
, ocp-indent
}:

buildDunePackage rec {
  pname = "jupyter";
  version = "2.8.2";

  minimumOCamlVersion = "4.10";
  duneVersion = "3";

  src = fetchurl {
    url = "https://github.com/akabe/ocaml-jupyter/archive/refs/tags/v${version}.tar.gz";
    sha256 = "sha256-npQcQyksrSknCHngqbORAqgwn2j3UFWFTVeixRiU+DQ=";
  };

  nativeBuildInputs = [
    cppo
  ];

  buildInputs = [
    base
    uuidm
    base64
    lwt
    lwt_ppx
    logs
    stdint
    zmq
    zmq-lwt
    yojson
    ppx_yojson_conv
    ppx_deriving
    cryptokit
  ];

  propagatedBuildInputs = [ opam ];

  checkInputs = [
    ounit2
    ocp-indent
  ];

  meta = {
    homepage = https://github.com/akabe/ocaml-jupyter;
    description = "An OCaml kernel for Jupyter (IPython) notebook";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ kwshi ];
    mainProgram = "ocaml-jupyter-opam-genspec";
  };
}
```

**Comment by kwshi** [comment]
I am still interested in this-- for the time being, I'm a bit busy, but I'll try to find a time to check/try this out in the next few weeks.

File: pkgs/development/ocaml-modules/jupyter/default.nix [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  pname = "jupyter";
+  version = "2.8.2";
+
+  minimumOCamlVersion = "4.10";
+  duneVersion = "3";
+
+  src = fetchurl {
+    url = "https://github.com/akabe/ocaml-jupyter/archive/refs/tags/v${version}.tar.gz";
+    sha256 = "sha256-npQcQyksrSknCHngqbORAqgwn2j3UFWFTVeixRiU+DQ=";
+  };
```
Comment (Janik-Haag) [inline pkgs/development/ocaml-modules/jupyter/default.nix:33]: this should be changed to the `fetchFromGitHub` wrapper

File: pkgs/development/ocaml-modules/jupyter/default.nix [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+, ounit2
+, ocp-indent
+}:
+
+buildDunePackage rec {
+  pname = "jupyter";
+  version = "2.8.2";
+
+  minimumOCamlVersion = "4.10";
+  duneVersion = "3";
```
Comment (Janik-Haag) [inline pkgs/development/ocaml-modules/jupyter/default.nix:28]: You probably want to update those versions.

**Review by Janik-Haag — CHANGES_REQUESTED** [review]
When trying to execute the build derivation it throws an error about opam, this might be fixed after version updates and a rebase. 
```
./result/bin/ocaml-jupyter-opam-genspec: line 3: opam: command not found
```

**Comment by Janik-Haag** [comment]
> I am still interested in this-- for the time being, I'm a bit busy, but I'll try to find a time to check/try this out in the next few weeks.

Take your time :D

**Comment by AndersonTorres** [comment]
Closing as dead.

Feel free to reopen.

**Pull request closed without merging**